driver.go

```diff
@@ -119,37 +206,93 @@ func (c *connector) Driver() driver.Driver {
 }
 
 type conn struct {
+	closed      bool
 	client      *spanner.Client
 	adminClient *adminapi.DatabaseAdminClient
-	roTx        *spanner.ReadOnlyTransaction
-	rwTx        *rwTx
-	name        string
+	tx          contextTransaction
+	database    string
+}
+
+// Ping implements the driver.Pinger interface.
+// returns ErrBadConn if the connection is no longer valid.
+func (c *conn) Ping(ctx context.Context) error {
+	if c.closed {
+		return driver.ErrBadConn
+	}
+	rows, err := c.QueryContext(ctx, "SELECT 1", []driver.NamedValue{})
+	if err != nil {
+		return driver.ErrBadConn
+	}
+	defer rows.Close()
+	values := make([]driver.Value, 1)
+	if err := rows.Next(values); err != nil {
+		return driver.ErrBadConn
+	}
+	if values[0] != int64(1) {
+		return driver.ErrBadConn
+	}
+	return nil
+}
+
+// ResetSession implements the driver.SessionResetter interface.
+// returns ErrBadConn if the connection is no longer valid.
+func (c *conn) ResetSession(_ context.Context) error {
+	if c.closed {
+		return driver.ErrBadConn
+	}
+	if c.inTransaction() {
+		if err := c.tx.Rollback(); err != nil {
+			return driver.ErrBadConn
+		}
+	}
+	return nil
+}
+
+// IsValid implements the driver.Validator interface.
+func (c *conn) IsValid() bool {
+	return !c.closed
 }
 
 func (c *conn) Prepare(query string) (driver.Stmt, error) {
-	panic("Using PrepareContext instead")
+	return c.PrepareContext(context.Background(), query)
 }
 
-func (c *conn) PrepareContext(ctx context.Context, query string) (driver.Stmt, error) {
-	// TODO(jbd): Mention emails need to be escaped.
```

Comment:
No longer needed, as the new parser is able to detect string literals and comments that may contain email addresses and other strings that could otherwise be seen as parameter names.